packages/cross-chain-replication-contracts/contracts/ZkCertificateRegistryReplica.sol

```diff
@@ -0,0 +1,209 @@
+
+    emit Initialized(_authorizedUpdater);
+  }
+
+  /**
+   * @notice Returns the current merkle root (last one in the array)
+   * @return The current merkle root
+   */
+  function merkleRoot() public view override returns (bytes32) {
+    if (merkleRoots.length == 0) {
+      return bytes32(0);
+    }
+    return merkleRoots[merkleRoots.length - 1];
+  }
+
+  /**
+   * @notice Returns the length of the merkle roots array
+   * @return The length of the merkle roots array
+   */
+  function merkleRootsLength() public view override returns (uint256) {
+    return merkleRoots.length;
+  }
+
+  /**
+   * @notice Returns a slice of the merkle roots array starting from _startIndex
+   * @param _startIndex The index to start returning roots from
+   * @return Array of merkle roots from the start index
+   */
+  function getMerkleRoots(
+    uint256 _startIndex
+  ) public view override returns (bytes32[] memory) {
+    require(_startIndex < merkleRoots.length, 'Start index out of bounds');
+
+    return getMerkleRoots(_startIndex, merkleRoots.length);
+  }
+
+  /**
+   * @notice Returns a slice of the merkle roots array starting from _startIndex and ending at _endIndex
+   * @param _startIndex The index to start returning roots from
+   * @param _endIndex The index to end returning roots at
+   * @return Array of merkle roots from the start index to the end index
+   */
+  function getMerkleRoots(
+    uint256 _startIndex,
+    uint256 _endIndex
+  ) public view override returns (bytes32[] memory) {
+    require(_startIndex < _endIndex, 'Start index must be less than end index');
+    require(_endIndex <= merkleRoots.length, 'End index out of bounds');
+
+    uint256 length = _endIndex - _startIndex;
+    bytes32[] memory roots = new bytes32[](length);
+
+    for (uint256 i = 0; i < length; i++) {
+      roots[i] = merkleRoots[_startIndex + i];
+    }
+
+    return roots;
+  }
+
+  /**
+   * @notice Verifies if a given merkle root is valid for proof verification
+   * @param _merkleRoot The merkle root to verify
+   * @return True if the merkle root is valid
+   */
+  function verifyMerkleRoot(
+    bytes32 _merkleRoot
+  ) public view override returns (bool) {
+    uint256 _merkleRootIndex = merkleRootIndex[_merkleRoot];
+    return _merkleRootIndex >= merkleRootValidIndex && _merkleRootIndex > 0;
+  }
+
+  /**
+   * @notice Updates the registry state - only callable by authorized updater
+   * @param newMerkleRoots Array of new merkle roots to append
+   * @param oldestValidMerkleRoot The merkle root that marks the validity boundary from the source registry
+   * @param newQueuePointer New queue pointer value
+   */
+  function updateState(
+    bytes32[] calldata newMerkleRoots,
+    bytes32 oldestValidMerkleRoot,
+    uint256 newQueuePointer
+  ) external onlyAuthorizedUpdater {
+    require(newMerkleRoots.length > 0, 'Must provide at least one merkle root');
+    require(
+      newQueuePointer > currentQueuePointer || newQueuePointer == 0,
```

Comment:
can it ever happen that the order of two update messages is switched? In this case this 'require' will invalidate the second message.